post/2017-10-30-how-test-serverless-applications.md

```diff
@@ -0,0 +1,95 @@
+---
+title: How to Test Serverless Applications
+description: "Tips from the Serverless team - how we test serverless applications."
+date: 2017-10-30
+thumbnail: 'https://s3-us-west-2.amazonaws.com/assets.blog.serverless.com/testing-apps.jpg'
+layout: Post
+authors:
+  - EslamHefnawy
+---
+
+Serverless applications are quickly gaining in complexity. Testing is key.
+
+I’ve been building the Serverless Framework for 2 years now, and during that time it's been my focus to create as smooth a testing and debugging experience as possible. In this article, I’ll share some techniques that you can use with the [Serverless Framework](https://www.serverless.com/framework) to test and debug your serverless application during development. 
+
+We'll focus on testing serverless functions, since this is where most of the development happens.
+
+# Unit testing
+
+This one is a gimme. You should always start with unit tests, whether or not your codebase is serverless. Forget about Lambda, the handlers, the events—just organize your codebase in an easy-to-test structure that follows your language’s best practice.
+
+Your handler should always be a thin layer that uses modules out of your code library. If those modules are well-covered with unit tests, then testing the serverless part of your application (i.e., the handlers) will be easy during the integration tests discussed below.
+
+# Integration testing
+
+Now that you’ve covered your codebase, it’s time to move on to your handlers with overall integration tests. Let's see how all of those units you’ve been testing individually work together.
+
+**Work with stages**<br>
+Since you’ll be interacting with the actual infrastructure pieces in your application, you'll need to make sure you stage your application during development. Set up a dev stage—the default when using the Serverless Framework—for all application infrastructure (databases, buckets, etc.) that your codebase will use during the integration tests.
+
+**Set up event mocks**<br>
+One other step that you’ll need to take before starting the tests is to have some event mocks prepared for all your handlers depending on one type of event the handler is expecting. For example, if your serverless function is subscribing to an S3 event source, make sure you have a JSON file that mocks that S3 event that AWS sends out. You can get that by trying it out yourself only once on AWS, and store it somewhere for future reference.
```

Comment:
What about a code snippet here to demonstrate? Show how to deploy a simple function that just logs the event, then trigger it once and read it from the logs. Then store in a JSON file.

Reply:
yep! added that 😊